swap/protocol.go

```diff
@@ -73,25 +73,38 @@ func (s *Swap) Stop() error {
 
 func (s *Swap) verifyHandshake(msg interface{}) error {
 	handshake, ok := msg.(*SwapHandshakeMsg)
-	var empty common.Address
-	if !ok || handshake.Beneficiary == empty {
+	if !ok || (handshake.ContractAddress == common.Address{}) {
 		return ErrEmptyAddressInSignature
 	}
+
+	err := s.verifyContract(context.TODO(), handshake.ContractAddress)
+
+	if err != nil {
+		log.Warn("swap handshake failed", err)
+		return err
+	}
+
 	return nil
 }
 
 func (s *Swap) run(p *p2p.Peer, rw p2p.MsgReadWriter) error {
 	protoPeer := protocols.NewPeer(p, rw, Spec)
 
 	answer, err := protoPeer.Handshake(context.TODO(), &SwapHandshakeMsg{
-		Beneficiary: s.owner.address,
+		ContractAddress: s.owner.Contract,
 	}, s.verifyHandshake)
 
 	if err != nil {
 		return err
 	}
 
-	swapPeer := NewPeer(protoPeer, s, s.backend, answer.(*SwapHandshakeMsg).Beneficiary)
+	beneficiary, err := s.getContractOwner(context.TODO(), answer.(*SwapHandshakeMsg).ContractAddress)
+
+	if err != nil {
+		return err
+	}
+
+	swapPeer := NewPeer(protoPeer, s, s.backend, beneficiary, answer.(*SwapHandshakeMsg).ContractAddress)
```

Comment:
map access below needs a lock, since it is shared among the concurrent protocol routines